From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 15:35:19 +0300
Subject: [PATCH] Inkwell config


diff --git a/.gitignore b/.gitignore
index 97e78e27ee0eea2c8b24886eeb19164d552323fe..d9c459d2cdafc4efea5bd7200acc540087d22426 100644
--- a/.gitignore
+++ b/.gitignore
@@ -38,3 +38,6 @@
 # vs code
 /.vscode
 /.factorypath
+
+#git
+*.rej
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d4c5b06fe4be177442ef1abaf1b223c6c2845930..1d0a65e1a76faf67d05e32bf074809239cdbbbb5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2257,6 +2257,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Inkwell start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getInkwellConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        @NotNull
+        public java.util.Properties getServerProperties() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Inkwell end
+
         /**
          * Sends the component to the player
          *
