From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 20 Jan 2024 19:05:28 +0300
Subject: [PATCH] Creative block noclip


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d048ae07cc33fd77d128cc1ebf88b0804969fa3c..cd4fd5a94eab5ca04a2148b2ac3376c7f4133bdf 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3752,4 +3752,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // Inkwell start
+    /**
+     * Returns is player in creative fly or spectator
+     *
+     * @return is player in creative fly or spectator
+     */
+    boolean isCreativeOrSpectator();
+    //Inkwell end
 }
