From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 20 Jan 2024 19:05:28 +0300
Subject: [PATCH] Creative noclip fly api


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ae61a39b25267b84fe0b8766e4b12d9b24b44ded..6740b11e71002d9327a5b51f0a28fa9afa4b45ce 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3574,4 +3574,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // Inkwell start
+    /**
+     * Returns is player in creative fly or spectator
+     *
+     * @return is player in creative fly or spectator
+     */
+    boolean isCreativeFlyOrSpectator();
+    //Inkwell end
 }
