From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 20 Jan 2024 19:05:28 +0300
Subject: [PATCH] Creative block noclip


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6c327a07bf8a6aa11a2d7dad12b2830acc539484..69bbb6697ee87d4e9380061d3057879399fced55 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3806,4 +3806,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // Inkwell start
+    /**
+     * Returns is player in creative fly or spectator
+     *
+     * @return is player in creative fly or spectator
+     */
+    boolean isCreativeOrSpectator();
+    //Inkwell end
 }
