From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 17:21:54 +0300
Subject: [PATCH] Firework sit player


diff --git a/src/main/java/org/bukkit/entity/Firework.java b/src/main/java/org/bukkit/entity/Firework.java
index 217d348ad0bbef720b25d3b507a55ca8105b7731..d76fe8e6b8cf446b1a1b659db2f1431a84148452 100644
--- a/src/main/java/org/bukkit/entity/Firework.java
+++ b/src/main/java/org/bukkit/entity/Firework.java
@@ -185,4 +185,11 @@ public interface Firework extends Projectile {
      */
     void setTicksToDetonate(int ticks);
     // Paper end
+
+    /**
+     * Sit player on firework.
+     *
+     * @param player player sit on firework.
+     */
+    void sitPlayer(Player player);
 }
