From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 20 Jan 2024 19:05:28 +0300
Subject: [PATCH] Creative block noclip


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 254a02ddb5dc867c9dd6c2086791f7ab94247fd3..714d532d5d49a27c8f74f78da8b69b25b41fef40 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3687,4 +3687,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // Inkwell start
+    /**
+     * Returns is player in creative fly or spectator
+     *
+     * @return is player in creative fly or spectator
+     */
+    boolean isCreativeOrSpectator();
+    //Inkwell end
 }
