From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <mia.krs3007@yandex.ru>
Date: Sun, 3 Dec 2023 20:02:19 +0300
Subject: [PATCH] Add server mod name


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 6776016211347c5f60fa8c6dc9a6b36f50162e15..f24857c05bf7c7bf007dd7c3a5774b2dc12b8d74 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -130,6 +130,11 @@ public class InkwellConfig {
         return config.getStringList(key);
     }
 
+    public static String serverModName;
+    private static void server() {
+        serverModName = getString("server.server-mod-name", "Inkwell", "Changes the server brand in f3 and server ping");
+    }
+
     public static boolean allowVoidTrading;
     private static void villager() {
         allowVoidTrading = getBoolean("entity.villager.allow-void-trading", false, "Allows you to interact with merchants in unloaded chunks");
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 0ab2183377d53ffd509720da579910fd9932ccbd..11aac3b66a7f1aa92033fdc38ffd50f2ea0c25de 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1697,7 +1697,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "Inkwell"; // Inkwell - Inkwell > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        return io.inkwellmc.inkwell.InkwellConfig.serverModName; // Inkwell - Inkwell > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
