From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 17:24:25 +0300
Subject: [PATCH] Firework sit player api


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 288910fb168ddc5d3a61971778b8038a56772fa8..d2180f301dae45a3194dfd42ec9f99a0b0b42a3a 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -197,7 +197,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         super.onHitEntity(entityHitResult);
         if (!this.level().isClientSide) {
             // CraftBukkit start
-            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled() && this.hasBeenShot) {
                 this.explode();
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
index 68c5af9b67a2834ee6e2f80ceefa19c3a982b8ed..6e3ec95d5dab54d773024daba8c855c50ee5d4de 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
@@ -173,6 +173,11 @@ public class CraftFirework extends CraftProjectile implements Firework {
         this.getHandle().lifetime = ticks;
     }
 
+    @Override
+    public void sitPlayer(@org.jetbrains.annotations.NotNull org.bukkit.entity.Player player) {
+        this.addPassenger(player);
+    }
+
     void applyFireworkEffect(FireworkMeta meta) {
         ItemStack item = this.getHandle().getItem();
         CraftItemStack.applyMetaToItem(item, meta);
