From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 17:24:25 +0300
Subject: [PATCH] Firework sit player api


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index b2f08889139dc447f7071f1c81456035bf8de31e..2f3d9d7265f22538191f2010137a9a6accb06748 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -197,7 +197,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         super.onHitEntity(entityHitResult);
         if (!this.level().isClientSide) {
             // CraftBukkit start
-            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled() && this.hasBeenShot) {
                 this.explode();
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
index c13cdbaf7abdf120a3969f8e887b4c3b78989c9c..cdef9d54934dd2edf7119aa55ec861c439455d12 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
@@ -177,6 +177,11 @@ public class CraftFirework extends CraftProjectile implements Firework {
         this.getHandle().lifetime = ticks;
     }
 
+    @Override
+    public void sitPlayer(org.bukkit.entity.Player player) {
+        this.addPassenger(player);
+    }
+
     void applyFireworkEffect(FireworkMeta meta) {
         ItemStack item = this.getHandle().getItem();
         CraftItemStack.applyMetaToItem(item, meta);
