From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 22:50:50 +0300
Subject: [PATCH] Disable out of order chat option


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 5eeb8ed0df4f3f5ab8e0f9f307b526d28e6ccd40..99f5e8a56d5d9eadfe82a4a4a393b862e5d6c4a8 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -227,4 +227,9 @@ public class InkwellConfig {
                 }, () -> MinecraftServer.LOGGER.warn("Unknown entity \"" + name + "\"")));
         setComment("dab", "Optimizes entity brains when", "they're far away from the player");
     }
+
+    public static boolean disableOutOfOrderChat;
+    private static void miscSettings() {
+        disableOutOfOrderChat = getBoolean("misc.disable-out-of-order-chat", true);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/network/chat/SignedMessageChain.java b/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
index 85a8a687b1568a56e3e646b37ef78b562c1b8a82..26f696783d9649dd7154c3a3fcaca939b4632742 100644
--- a/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
+++ b/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
@@ -38,7 +38,7 @@ public class SignedMessageChain {
                 throw new SignedMessageChain.DecodeException(Component.translatable("chat.disabled.chain_broken"), false); // Paper - diff on change (if disconnects, need a new kick event cause)
             } else if (playerPublicKey.data().hasExpired()) {
                 throw new SignedMessageChain.DecodeException(Component.translatable("chat.disabled.expiredProfileKey", org.bukkit.event.player.PlayerKickEvent.Cause.EXPIRED_PROFILE_PUBLIC_KEY), false); // Paper - kick event causes
-            } else if (body.timeStamp().isBefore(this.lastTimeStamp)) {
+            } else if (!io.inkwellmc.inkwell.InkwellConfig.disableOutOfOrderChat && body.timeStamp().isBefore(this.lastTimeStamp)) { // Inkwell
                 throw new SignedMessageChain.DecodeException(Component.translatable("multiplayer.disconnect.out_of_order_chat", org.bukkit.event.player.PlayerKickEvent.Cause.OUT_OF_ORDER_CHAT), true); // Paper - kick event causes
             } else {
                 this.lastTimeStamp = body.timeStamp();
