From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sat, 30 Dec 2023 22:50:50 +0300
Subject: [PATCH] Disable out of order chat option


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index ae201beb80e3fb74243970f7d42d0939a812147c..6c83e8f5cef07b6daf8ef10bffaff9ad64290e7c 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -227,4 +227,9 @@ public class InkwellConfig {
                 }, () -> MinecraftServer.LOGGER.warn("Unknown entity \"" + name + "\"")));
         setComment("dab", "Optimizes entity brains when", "they're far away from the player");
     }
+
+    public static boolean disableOutOfOrderChat;
+    private static void miscSettings() {
+        disableOutOfOrderChat = getBoolean("misc.disable-out-of-order-chat", true);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 3c0651fa5a5db880202c9a3805a6455269c5f16d..8e50a23f9064b3616d9a147d753fccd0e63af98a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2420,6 +2420,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     }
 
     private boolean updateChatOrder(Instant timestamp) {
+        if (io.inkwellmc.inkwell.InkwellConfig.disableOutOfOrderChat) return true;
         Instant instant1;
 
         do {
