From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Wed, 17 Jan 2024 14:57:29 +0300
Subject: [PATCH] Villager lobotomize


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 0fd70a9ed7f82220b1dd917c97a36e612854b511..2fc4419f3501bc3b4ec3929e99029f99d5b43f72 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -141,12 +141,23 @@ public class InkwellConfig {
     }
 
     public static boolean allowVoidTrading;
+    public static boolean villagerLobotomizeEnabled = false;
+    public static int villagerLobotomizeCheckInterval = 100;
     private static void villager() {
         allowVoidTrading = getBoolean(
                 "entity.villager.allow-void-trading",
                 false,
                 "Allows you to interact with merchants in unloaded chunks"
         );
+
+        villagerLobotomizeEnabled = getBoolean(
+                "entity.villager.lobotomize.enabled",
+                villagerLobotomizeEnabled
+        );
+        villagerLobotomizeCheckInterval = getInt(
+                "entity.villager.lobotomize.check-interval",
+                villagerLobotomizeCheckInterval
+        );
     }
 
     public static boolean usePigmanSpawnLimit;
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 6339967b154af1ca9632c7b8b315d612f4b80ea2..1c2beaaeba8e79fe47e59c76530e69e75789198e 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -143,6 +143,9 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         return holder.is(PoiTypes.MEETING);
     });
 
+    private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Inkwell
+    private int notLobotomizedCount = 0; // Inkwell
+
     public Villager(EntityType<? extends Villager> entityType, Level world) {
         this(entityType, world, VillagerType.PLAINS);
     }
@@ -207,6 +210,44 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         brain.updateActivityFromSchedule(this.level().getLevelData().getDayTime(), this.level().getLevelData().getGameTime()); // Folia - region threading
     }
 
+    private boolean checkLobotomized() {
+        int interval = io.inkwellmc.inkwell.InkwellConfig.villagerLobotomizeCheckInterval;
+        if (this.notLobotomizedCount > 3) {
+            // check half as often if not lobotomized for the last 3+ consecutive checks
+            interval *= 2;
+        }
+        if (this.level().getGameTime() % interval == 0) {
+            // offset Y for short blocks like dirt_path/farmland
+            this.isLobotomized = !canTravelFrom(BlockPos.containing(this.position().x, this.getBoundingBox().minY + 0.0625D, this.position().z));
+            if (this.isLobotomized) {
+                this.notLobotomizedCount = 0;
+            } else {
+                this.notLobotomizedCount++;
+            }
+        }
+        return this.isLobotomized;
+    }
+    private boolean canTravelFrom(BlockPos pos) {
+        return canTravelTo(pos.east()) || canTravelTo(pos.west()) || canTravelTo(pos.north()) || canTravelTo(pos.south());
+    }
+    private boolean canTravelTo(BlockPos pos) {
+        net.minecraft.world.level.block.state.BlockState state = this.level().getBlockStateIfLoaded(pos);
+        if (state == null) {
+            // chunk not loaded
+            return false;
+        }
+        net.minecraft.world.level.block.Block bottom = state.getBlock();
+        if (bottom instanceof net.minecraft.world.level.block.FenceBlock ||
+                bottom instanceof net.minecraft.world.level.block.FenceGateBlock ||
+                bottom instanceof net.minecraft.world.level.block.WallBlock) {
+            // bottom block is too tall to get over
+            return false;
+        }
+        net.minecraft.world.level.block.Block top = level().getBlockState(pos.above()).getBlock();
+        // only if both blocks have no collision
+        return !bottom.hasCollision && !top.hasCollision;
+    }
+
     @Override
     protected void ageBoundaryReached() {
         super.ageBoundaryReached();
@@ -252,10 +293,19 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         // Paper start
         this.customServerAiStep(false);
     }
-    protected void customServerAiStep(final boolean inactive) {
+    protected void customServerAiStep(boolean inactive) { // Inkwell - not final
         // Paper end
         this.level().getProfiler().push("villagerBrain");
+        // Inkwell start
+        if (io.inkwellmc.inkwell.InkwellConfig.villagerLobotomizeEnabled) {
+            // treat as inactive if lobotomized
+            inactive = inactive || checkLobotomized();
+        } else {
+            this.isLobotomized = false;
+        }
         if (!inactive) this.getBrain().tick((ServerLevel) this.level(), this); // Paper
+        else if (this.isLobotomized && shouldRestock()) restock();
+        // Inkwell end
         this.level().getProfiler().pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
index 771eb799b032fcf4fc33c5e445b7fa252baee72b..1bcf5d8ef53534fd06b60b1eef576bd306a14f38 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -260,4 +260,11 @@ public class CraftVillager extends CraftAbstractVillager implements Villager {
         getHandle().getGossips().gossips.clear();
     }
     // Paper end
+
+    // Inkwell start
+    @Override
+    public boolean isLobotomized() {
+        return getHandle().isLobotomized();
+    }
+    // Inkwell end
 }
