From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sun, 21 Jan 2024 13:52:11 +0300
Subject: [PATCH] Snowball knockback player


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index ce52e5f41d892e5b7605411bb55ab11319f7e683..318648eb7fce2687e5b7af5a8d586ac9e9017efd 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -199,9 +199,11 @@ public class InkwellConfig {
 
     public static boolean disableOutOfOrderChat;
     public static int pistonPushLimit;
+    public static boolean snowballKnockbackPlayer;
     private static void miscSettings() {
         disableOutOfOrderChat = getBoolean("misc.disable-out-of-order-chat", true);
         pistonPushLimit = getInt("misc.piston-push-limit", 12, "The number of blocks the piston can move. The default is 12.");
+        snowballKnockbackPlayer = getBoolean("misc.snowball-knockback-player", false, "Value defines can snowball kockback player");
     }
 
     public static Map<String, Integer> projectileTimeouts;
diff --git a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
index 2b4d206c0d31ba38d7b2af654bd420e85145d441..83de28a457d943fd7d2d8a1c98d398891dc477e9 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
@@ -61,6 +61,13 @@ public class Snowball extends ThrowableItemProjectile {
         int i = entity instanceof Blaze ? 3 : 0;
 
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), (float) i);
+
+        // Inkwell start - make snowball can knockback player
+        if (io.inkwellmc.inkwell.InkwellConfig.snowballKnockbackPlayer && entity instanceof net.minecraft.server.level.ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((net.minecraft.server.level.ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ(), this, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.DAMAGE);
+        }
+        // Inkwell end - make snowball can knockback player
     }
 
     @Override
