From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sun, 21 Jan 2024 13:57:42 +0300
Subject: [PATCH] Spectator advancements


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 02b9fcff7f3cdcca809082dff879a8d87c4e664c..492d4cc62f3a51b736e7c3a59c90956513c2caca 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -287,7 +287,9 @@ public class InkwellConfig {
     }
 
     public static boolean playerCreativeBlockNoclip;
+    public static boolean spectatorCanGetAdvancement;
     private static void playerSettings() {
         playerCreativeBlockNoclip = getBoolean("player.creative-block-noclip", false, "Allows noclip blocks in creative");
+        spectatorCanGetAdvancement = getBoolean("player.spectator-advancements", true, "Value defines spectator can get advancements");
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index acc49f66bf34e2507d0ee6fec0a56b11bfc68f46..e8cd53fac8c70227b1528b6ac8c856447e6668bb 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -224,6 +224,12 @@ public class PlayerAdvancements {
     }
 
     public boolean award(Advancement advancement, String criterionName) {
+        // Inkwell start - spectator don't get advancement
+        if (!io.inkwellmc.inkwell.InkwellConfig.spectatorCanGetAdvancement && player.gameMode.getGameModeForPlayer() == net.minecraft.world.level.GameType.SPECTATOR) {
+            return false;
+        }
+        // Inkwell end - spectator don't get advancement
+
         boolean flag = false;
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
         boolean flag1 = advancementprogress.isDone();
diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java.rej b/src/main/java/net/minecraft/server/PlayerAdvancements.java.rej
new file mode 100644
index 0000000000000000000000000000000000000000..4b6cf8a58701076a9935ad3821f8cbd410d06be5
--- /dev/null
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java.rej
@@ -0,0 +1,14 @@
+diff a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java	(rejected hunks)
+@@ -221,6 +221,12 @@ public class PlayerAdvancements {
+     }
+ 
+     public boolean award(AdvancementHolder advancement, String criterionName) {
++        // Inkwell start - spectator don't get advancement
++        if (!io.inkwellmc.inkwell.InkwellConfig.spectatorCanGetAdvancement && player.gameMode.getGameModeForPlayer() == net.minecraft.world.level.GameType.SPECTATOR) {
++            return false;
++        }
++        // Inkwell end - spectator don't get advancement
++
+         boolean flag = false;
+         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
+         boolean flag1 = advancementprogress.isDone();
