From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sun, 17 Mar 2024 20:07:24 +0300
Subject: [PATCH] Option for obsidian pillars regen


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 1e9c6dd11d284f563fe1415b1b6d464c2e575aae..3b4f5290ea5e57c229982d98ef22ce1cd6d0d5ec 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -181,8 +181,10 @@ public class InkwellConfig {
     }
 
     public static boolean enderdragonAllowGriefing;
+    public static boolean regenObisidianPillarsWithEnderdragon;
     private static void enderdragonSettings() {
         enderdragonAllowGriefing = getBoolean("entity.enderdragon.allow-griefing", true);
+        regenObisidianPillarsWithEnderdragon = getBoolean("entity.enderdragon.regen-obsidian-pillars", true);
     }
 
     public static boolean creeperAllowGriefing;
diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/SpikeFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/SpikeFeature.java
index 260c3a7dc592fba220ad4a7febb43ee2c9279115..91d7a7ab016b1fc1d7c43c1dca68502772d4909f 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/SpikeFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/SpikeFeature.java
@@ -41,7 +41,7 @@ public class SpikeFeature extends Feature<SpikeConfiguration> {
     public static List<SpikeFeature.EndSpike> getSpikesForLevel(WorldGenLevel world) {
         RandomSource randomSource = RandomSource.create(world.getSeed());
         long l = randomSource.nextLong() & 65535L;
-        return SPIKE_CACHE.getUnchecked(l);
+        return io.inkwellmc.inkwell.InkwellConfig.regenObisidianPillarsWithEnderdragon ? SPIKE_CACHE.getUnchecked(l) : java.util.Collections.emptyList();
     }
 
     @Override
