From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Mon, 25 Mar 2024 18:23:16 +0300
Subject: [PATCH] Option for vanilla end teleportation


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 7fe175c4faf229a45f8254d500814392cb98a9cc..3158d69ece3a4bd3905b53c70b75d3f76d681473 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -245,10 +245,12 @@ public class InkwellConfig {
     public static boolean disableOutOfOrderChat;
     public static int pistonPushLimit;
     public static boolean snowballKnockbackPlayer;
+    public static boolean useVanillaEndTeleportation = false;
     private static void miscSettings() {
         disableOutOfOrderChat = getBoolean("misc.disable-out-of-order-chat", true);
         pistonPushLimit = getInt("misc.piston-push-limit", 12, "The number of blocks the piston can move. The default is 12.");
         snowballKnockbackPlayer = getBoolean("misc.snowball-knockback-player", false, "Value defines can snowball kockback player");
+        useVanillaEndTeleportation = getBoolean("misc.vanilla-end-teleportation", useVanillaEndTeleportation, "This can fix wither platform obsidian farm");
     }
 
     public static Map<String, Integer> projectileTimeouts;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 325d9e34d77ea29a2cb08a2b8af6acd756d7e4f7..cb7d86f45ec3dc1044cb5163ddb378dddcd33799 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4165,9 +4165,17 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
 
                             // the portal obsidian is placed at targetPos.y - 2, so if we want to place the entity
                             // on the obsidian, we need to spawn at targetPos.y - 1
-                            portalInfoCompletable.complete(
-                                new PortalInfo(Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, 90.0f, 0.0f, destination, null)
-                            );
+                            // Inkwell start
+                            if (io.inkwellmc.inkwell.InkwellConfig.useVanillaEndTeleportation) {
+                                Vec3 finalPos = this instanceof Player ? Vec3.atBottomCenterOf(targetPos.below()) : Vec3.atBottomCenterOf(targetPos);
+                                portalInfoCompletable.complete(
+                                    new PortalInfo(finalPos, this.getDeltaMovement(), 90.0f, 0.0f, destination, null)
+                                );
+                            } else {
+                                portalInfoCompletable.complete(
+                                    new PortalInfo(Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, 90.0f, 0.0f, destination, null)
+                                );
+                            } // Inkwell end
                         }
                     );
                 } else {
