From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sun, 21 Jan 2024 13:52:11 +0300
Subject: [PATCH] Snowball knockback player


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 75c4db957c46f54e0c14dc41a9cbebbdfc897484..9d4c5fb5c50860942ef0e0646d476123bfd73cab 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -227,9 +227,11 @@ public class InkwellConfig {
 
     public static boolean disableOutOfOrderChat;
     public static int pistonPushLimit;
+    public static boolean snowballKnockbackPlayer;
     private static void miscSettings() {
         disableOutOfOrderChat = getBoolean("misc.disable-out-of-order-chat", true);
         pistonPushLimit = getInt("misc.piston-push-limit", 12, "The number of blocks the piston can move. The default is 12.");
+        snowballKnockbackPlayer = getBoolean("misc.snowball-knockback-player", false, "Value defines can snowball kockback player");
     }
 
     public static Map<String, Integer> projectileTimeouts;
diff --git a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
index 718e120c9768cf716b32d3d652f53f1dda925168..7a40f38b745964f461f9d21021951685b1bf70b0 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
@@ -55,6 +55,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = entityHitResult.getEntity();
         int i = entity instanceof Blaze ? 3 : 0;
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), (float)i);
+        // Inkwell start - make snowball can knockback player
+        if (io.inkwellmc.inkwell.InkwellConfig.snowballKnockbackPlayer && entity instanceof net.minecraft.server.level.ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((net.minecraft.server.level.ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ(), this);
+        }
+        // Inkwell end - make snowball can knockback player
     }
 
     @Override
