From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sub-kek <subbkek@gmail.com>
Date: Sun, 21 Jan 2024 23:05:12 +0300
Subject: [PATCH] Griefing control


diff --git a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
index 536f6d01764940918613ddfb5c2af05f1fdcabb5..0bc76e6b84b337a287ef0af5625929169017300a 100644
--- a/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
+++ b/src/main/java/io/inkwellmc/inkwell/InkwellConfig.java
@@ -185,6 +185,16 @@ public class InkwellConfig {
         enderdragonAllowGriefing = getBoolean("entity.enderdragon.allow-griefing", true);
     }
 
+    public static boolean creeperAllowGriefing;
+    private static void creeperSettings() {
+        creeperAllowGriefing = getBoolean("entity.creeper.allow-griefing", true);
+    }
+
+    public static boolean endermanAllowGriefing;
+    private static void endermanSettings() {
+        endermanAllowGriefing = getBoolean("entity.enderman.allow-griefing", true);
+    }
+
     public static boolean allowSignColors;
     private static void blockEntity() {
         allowSignColors = getBoolean(
diff --git a/src/main/java/net/minecraft/world/entity/monster/Creeper.java b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
index 6a91960f93e624a4b975c1cf76698eb45ceda838..7b4c53ac11827dd0e5141752656e8378be0283d0 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Creeper.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
@@ -272,7 +272,7 @@ public class Creeper extends Monster implements PowerableMob {
             if (!event.isCancelled()) {
             // CraftBukkit end
             this.dead = true;
-            this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB); // CraftBukkit
+                this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), this.level().getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING) && io.inkwellmc.inkwell.InkwellConfig.creeperAllowGriefing ? Level.ExplosionInteraction.MOB : Level.ExplosionInteraction.NONE); // CraftBukkit // Inkwell
             this.discard();
             this.spawnLingeringCloud();
             // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
index bb4371bf31de61b9b1c7392d4187e712df671860..7698f4805cd7502cf4b4ecf5a087d8108a991150 100644
--- a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
+++ b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
@@ -522,6 +522,7 @@ public class EnderMan extends Monster implements NeutralMob {
 
         @Override
         public boolean canUse() {
+            if (!io.inkwellmc.inkwell.InkwellConfig.endermanAllowGriefing) return false; // Inkwell
             return this.enderman.getCarriedBlock() == null ? false : (!this.enderman.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? false : this.enderman.getRandom().nextInt(reducedTickDelay(2000)) == 0);
         }
 
@@ -567,6 +568,7 @@ public class EnderMan extends Monster implements NeutralMob {
 
         @Override
         public boolean canUse() {
+            if (!io.inkwellmc.inkwell.InkwellConfig.endermanAllowGriefing) return false; // Inkwell
             return this.enderman.getCarriedBlock() != null ? false : (!this.enderman.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? false : this.enderman.getRandom().nextInt(reducedTickDelay(20)) == 0);
         }
 
